packages/bff/src/services/purposeService.ts

```diff
@@ -1,54 +1,398 @@
-import { WithLogger } from "pagopa-interop-commons";
+import { WithLogger, toSetToArray } from "pagopa-interop-commons";
 import {
   PurposeId,
   PurposeVersionId,
   unsafeBrandId,
 } from "pagopa-interop-models";
-import { bffApi } from "pagopa-interop-api-clients";
-import { PagoPAInteropBeClients } from "../providers/clientProvider.js";
-import { purposeNotFound } from "../model/domain/errors.js";
-import { BffAppContext } from "../utilities/context.js";
+import { bffApi, purposeApi, tenantApi } from "pagopa-interop-api-clients";
+import {
+  AgreementProcessClient,
+  AuthorizationProcessClient,
+  CatalogProcessClient,
+  PagoPAInteropBeClients,
+  PurposeProcessClient,
+  TenantProcessClient,
+} from "../providers/clientProvider.js";
+import {
+  agreementNotFound,
+  eserviceDescriptorNotFound,
+  eServiceNotFound,
+  purposeDraftVersionNotFound,
+  purposeNotFound,
+  tenantNotFound,
+} from "../model/domain/errors.js";
+import { BffAppContext, Headers } from "../utilities/context.js";
+import { toBffApiCompactClient } from "../model/domain/apiConverter.js";
+import { isUpgradable } from "../model/modelMappingUtils.js";
+import { getLatestAgreement } from "./agreementService.js";
+import { getAllClients } from "./clientService.js";
+
+export const getCurrentVersion = (
+  purposeVersions: purposeApi.PurposeVersion[]
+): purposeApi.PurposeVersion | undefined => {
+  const statesToExclude: purposeApi.PurposeVersionState[] = [
+    purposeApi.PurposeVersionState.Values.WAITING_FOR_APPROVAL,
+    purposeApi.PurposeVersionState.Values.REJECTED,
+  ];
+  return purposeVersions
+    .filter((v) => !statesToExclude.includes(v.state))
+    .sort(
+      (a, b) =>
+        new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
+    )
+    .at(-1);
+};
+
+// eslint-disable-next-line max-params
+async function getPurposes(
```

Comment:
I need this function outside for future. Here for example it's used in two places https://github.com/pagopa/interop-be-monorepo/pull/862/files

Reply:
> I need this function outside for future. Here for example it's used in two places https://github.com/pagopa/interop-be-monorepo/pull/862/files

I saw the changes but, even there, since we use it both times inside `purposeServiceBuilder`, `getPurposes` and `enhancePurpose` could be defined inside it saving us a lot of parameters to be passed.

It will look like this:
```
export function purposeServiceBuilder(...){
  async function enhancePurpose(...){...}
  async function getPurposes(...){...}

  return {
    async ...
  }
}
```
What do you think?